### ISSUE TITLE
Does this support adblock bypass?

### ISSUE THREAD

**Issue by tjosm** [cazala/coin-hive-stratum#60]
I can't find anything about it mentioned. Can coinhive.min.js be saved and renamed?

**Comment by sunk818** [comment]
I have saved and renamed `https://coinhive.com/lib/coinhive.min.js` to my own web server. It also required me to edit the coinhive.min.js file. Mind your `http://` vs `https://`. `WEBSOCKET_SHARDS` also has `ws` for non-SSL and `wss` for SSL which points to where the coin-hive-stratum proxy is hosted.

Some malware detectors may check for keywords like `coinhive` or even block WebAssembly (wasm), which I'm not sure you can bypass easily.

**Comment by CineXMike** [comment]
You need to obfuscate the orignal file,
Then you will be bypassing ad-blocker,
For my example i use AdMiner.Anonymous();
Instead of CoinHive.Anonymous();

**Comment by cazala** [comment]
You can avoid AdBlock, you can follow [this guide](https://github.com/cazala/coin-hive-stratum/wiki/Deploy-to-now.sh-and-GitHub-Pages) and deploy the proxy to now.sh and serve the assets from github pages, or just serve [these assets](https://github.com/cazala/coin-hive-stratum/tree/gh-pages) from your own server. The code has been obfuscated and encoded, and the `CoinHive` global element can be accessed as `CH`.

**Comment by phr34k0** [comment]
Adblock now blocks now.sh

Possible to deploy to our own servers like before with ssl secure websocket? Ngrok is a pretty nice tool for localhost tunneling if we lazy to keep making subdomains or get a cheap 0.48 cents domain on Namecheap.

**Comment by cazala** [comment]
I ll soon add a guide on how to deploy the proxy to a digital ocean droplet ($5/m, and you can get $50 dollars in credit for free with a github student pack) and an install.sh script to make it a single command install. Yesterday I setup a proxy up and running in like 5min, and it only takes another 5 min to buy a namecheap domain and point it to it or add a subdomain in cloudflare point it to it.

**Comment by phr34k0** [comment]
@cazala Managed to deploy a proxy on my own server with letsencrypt ssl websocket like you said, takes 5 mins. Possible to connect to coinhive.com using coin-hive-stratum? Still had coins in there

**Comment by cazala** [comment]
awesome (: no, coin-hive-stratum can't connect to coinhive.com, it was possible thru coin-hive-party but i didn't add that functionality to this package (to be honest it was a dirty hack).

**Comment by sunk818** [comment]
Couldn't you just use regular coin hive JavaScript for coinhive? Why all this round about?

**Comment by cazala** [comment]
coin-hive-party allowed connecting to coimhive.com's pool for adblock purposes

**Comment by phr34k0** [comment]
@sunk818 Because, adblock and anti viruses. And coinhive made some changes on their end that keeps throwing disconnection issues to websocket proxy server. coin-hive-party didn't help due to now.sh being blocked by adblock and restricted by now.sh service. Best is to just deploy your own proxy server with ssl capabilities.

**Comment by cp-fun-time** [comment]
What would be a good method of loading the script async and then running the miner after it loads? This way at least it can be obfuscated completely using a JS obfuscator, which will make it a bit more difficult for someone reading the page source to notice that it's a miner. I.e. the pool url is in the `script src=`, the attributes for the function look pretty miner-y

**Comment by cazala** [comment]
the assets/scripts are already encoded + obfuscated: https://github.com/cazala/coin-hive-stratum/wiki/Deploy-to-now.sh-and-GitHub-Pages#assets

**Comment by cazala** [comment]
you can find them here: https://github.com/cazala/coin-hive-stratum/tree/gh-pages

**Comment by tjosm** [comment]
I've tested those assets and they are working over http but gets blocked over https due to CORS policy. Would be nice to have it working over https.

I didn't test it on same server as my website.

**Comment by procloud** [comment]
you will have to run it native with a cert to avoid cors, or use http

**Comment by phr34k0** [comment]
I've tested it and had it working across multiple domains with https. Also added these in my Apache httpd.conf and restarted server after changes.

```
    #
    # Controls who can get stuff from this server.
    #
    Require all granted
    Header set Access-Control-Allow-Origin "*"
	# Header always append X-Frame-Options SAMEORIGIN
```

**Comment by phr34k0** [comment]
![image](https://user-images.githubusercontent.com/13272249/33651208-f6eb706c-da9f-11e7-8364-3316c1ce3ef6.png)

Added that TLS reject too

**Comment by cazala** [comment]
> I've tested those assets and they are working over http but gets blocked over https due to CORS policy. Would be nice to have it working over https.
I didn't test it on same server as my website.

If you fork the repo you can access the assets via https, ie: `https://tjosm.github.io/coin-hive-stratum/m.js` (you should change the name of the repo tho, so it doesn't say coin-hive in the url

**Comment by LDubya** [comment]
Hey @cazala, thanks for the info. I'm trying to find a solution to the following, and am not seeing it anywhere: let's say I just want to use coinhive's service (so no custom pool or anything) without adblock knocking it down. I figure I can download a copy of coinhive.min.js, rename all string instances of "coinhive" (including the filename), onfuscate, and serve this new file directly from my own server. Would that be enough? Or would not using a proxy lead it to being blocked?

**Comment by cazala** [comment]
the miner will still try to connect to one of coinhive's proxies (`wss://wsXX.coinhive.com/proxy`) and will get blocked by AdBlock most likely.

**Comment by LDubya** [comment]
@cazala @sunk818, Gotcha. The problem with the opt-in version is that this is for a cloud-based subscription service that organizations will sign up for and give their employees access to. If an employee has been given access to our software, the organization has already opted in, as those are the terms of use. But of course, understandably, employees use adblock on their office machines for security reasons. Sadly ad-blockers have decided on a blanket ban of browser-based mining, whether or not they are being used maliciously. And the explicit opt-in version would not work in this scenario either, as even having access to the service means that your organization has already opted in, and giving employees the option to opt out would undermine the agreement and disrupt the user experience.

I think this is a different use case than what the ad-blockers are addressing with their blanket ban of automatic mining. 

I think this is a good place to ask this because this repo focuses on a proxy solution, which we also think would be the only good solution. Coinhive's backend is simple enough to use for our needs, so we'd like to use their service.

Is there a way to set up a proxy using this repo, yet sill utilize coinhive's service instead of connecting to a custom pool, etc?

**Comment by sunk818** [comment]
@cazala will have definitive answer, but coin-hive-stratum is way to connect to stratum pools not coinhive pool. If you want to connect to coinhive pool, just use the javascript as it is. You should try, but I don't think adblockers block the websocket connections the pool, just the javascript. So, if you change the javascript (e.g. CoinHive text to CH), you should be different enough to bypass the adblockers yet connect to coinhive pools.

The whole point of this proxy is to bypass the 30% fee they charge. From a financial point, don't see why you'd want to give them 30% when there are better alternatives this coin-hive-stratum provides.

**Comment by cazala** [comment]
CoinHive miners connect to their proxies using websockets. The websocket upgrade request gets blocked by AdBlock.

coin-hive-stratum doesn't support your usecase as it is now, but it shouldn't be hard to add support for it (I've already done that for coin-hive-party).

stayed tuned and I might add it when i find some time. if anyone else ready this want the feature also let me know, if there's interest in it i will definitely add it.

**Comment by adamdupuis** [comment]
> the assets/scripts are already encoded + obfuscated: https://github.com/cazala/coin-hive-stratum/wiki/Deploy-to-now.sh-and-GitHub-Pages#assets

@cazala I found that Kaspersky detected the standard m.js file so I unobfuscated it as much as I could and hardcoded my proxy. After removing some console.log calls and obfuscating it a bit more I was able to bypass it undetected. I was wondering if you would provide your unobfuscated version of m.js so I could make some more changes?

Thanks!

**Comment by cazala** [comment]
@adamdupuis are you talking about: 

```
console.log("libUrl", libUrl);
console.log("proxyUrl", proxyUrl);
```

or what parts did you remove?

btw, the guide that [I promised here](https://github.com/cazala/coin-hive-stratum/issues/60#issuecomment-348162715) is out there: [Deploy to Digital Ocean](https://github.com/cazala/coin-hive-stratum/wiki/Deploy-to-Digital-Ocean)

**Comment by adamdupuis** [comment]
@cazala Yes those were the only two lines I removed aside from hardcoding the proxy so it wouldn't be required to include in plain text.

**Comment by adamdupuis** [comment]
Are you considering releasing the javascript source unobfuscated? Maybe in a new branch? I could waste more time sorting through it myself, but you know how programmers are ;P

**Comment by cazala** [comment]
The unobfuscated code is just the original miner, with a piece of code at the beginning that extracts the url from the url and puts it on the websocket shards array (which, in your case, would be removed), nothing else.

**Comment by Akuka** [comment]
@cazala Out of curiosity, can you tell us which tools you used for the obfuscation?

**Comment by cazala** [comment]
That's because their build script just pasted the that CoinHive.CONFIG object definition at the beginning of all their assets. The websocket shards are not even used in from the WebWorker (I don't even know if you can open a websocket from a WebWorker? I just stripped the whole thing off.

**Comment by cazala** [comment]
Copy + Paste here https://javascriptobfuscator.com/, so professional :P

**Comment by adamdupuis** [comment]
So I can just get rid of `WEBSOCKET_SHARDS`, `CAPTCHA_URL`, `MINER_URL` and `AUTH_URL` and on the last line hardcode my proxy in there?

**Comment by cazala** [comment]
In the Blob, yes. The only thing used is the LIB_URL.

**Comment by adamdupuis** [comment]
I think I've got it mostly finished, but when I try using the new code I get an error in the console:
`failed to asynchronously prepare wasm: TypeError: NetworkError when attempting to fetch resource.`

I've noticed that the CoinHive code references some files that you have not included and are not in CoinHive's /lib/ folder such as `cryptonight.temp.asm.js` and `cryptonight.wast` and I think that may have something to do with it. Any ideas @cazala?

Thanks again!

**Comment by phr34k0** [comment]
cryptonight.wast ? o.O

**Comment by adamdupuis** [comment]
Nevermind! Sorted it out as mentioned above.

**Comment by phr34k0** [comment]
XD Code fatigues

**Comment by cazala** [comment]
Can I ask what pieces your removed to workaround Kaspersky?

**Comment by adamdupuis** [comment]
You mean from your m.js? Just the two console.log() calls as mentioned above. Hardcoding the proxy didn't hurt either. I think it would be a good idea to minify the js as well as obfuscating it. I actually installed Kaspersky and kept trying different things until it didn't detect it. After all that I still have to call the script with src m.js proxy= for it to work.

**Comment by cazala** [comment]
okay, thanks for the info :)

**Comment by cazala** [comment]
@adamdupuis did you manage to make it work? If you have some info you could share it would be very helpful (: check this issue #100 thanks!

**Comment by adamdupuis** [comment]
Yes I did, as far as I know it is undetected after starting with the CoinHive's js file, but I haven't actually tested it against Kaspersky since.

**Comment by cazala** [comment]
Did you only change the `.js` or also did you rename or do something around the other assets (wasm, mem, asmjs)?

**Comment by adamdupuis** [comment]
I only hardcoded the proxy and removed the console.log() calls in m.js - no renaming of any files.

**Comment by phr34k0** [comment]
Looks like some browsers got smarter at blocking this https://github.com/keraf/NoCoin/issues/49 , https://github.com/stevespringett/disable-webassembly

**Comment by phr34k0** [comment]
Looks like webbrowsers got smarter at blocking this https://github.com/stevespringett/disable-webassembly

**Comment by phr34k0** [comment]
Yes that's what stated on coinhive site. 

> var hashesPerSecond = parseInt(miner.getHashesPerSecond());
> var totalHashes = parseInt(miner.getTotalHashes());
> var acceptedHashes = parseInt(miner.getAcceptedHashes());
> var numThreads = parseInt(miner.getNumThreads());

**Comment by unelink** [comment]
@cazala Hi! Any updates on this? https://github.com/cazala/coin-hive-stratum/issues/60#issuecomment-350055685 Do you think you will implement it anytime soon? Thank you!

**Comment by cazala** [comment]
not before the end of the year

**Comment by ghost** [comment]
Hi @cazala, any updates on this? https://github.com/cazala/coin-hive-stratum/issues/60#issuecomment-350055685 

Cheers!
